File: 08/python/homology.py
"""
Simplicial homology groups.

H_n(K) = ker(∂_n) / im(∂_{n+1})

The computation uses Smith normal form:

    Given boundary matrices ∂_n and ∂_{n+1}:
    - rank(∂_n) = r_n  (from Smith factors of ∂_n)
    - rank(∂_{n+1}) = r_{n+1}
    - number of n-simplices = k_n

    Then:
        β_n = k_n - r_n - r_{n+1}          (Betti number: rank of free part)
        torsion: Z/d_iZ for each invariant factor d_i > 1 of ∂_{n+1}

The Betti number β_n counts independent n-dimensional holes.
Torsion arises in non-orientable spaces (real projective plane, Klein bottle).
"""

from dataclasses import dataclass, field
from typing import List
from boundary import boundary_matrix
from smith import smith_factors, rank as matrix_rank


@dataclass
class HomologyGroup:
    """
    Represents H_n = Z^β ⊕ Z/d_1Z ⊕ ... ⊕ Z/d_kZ.

    betti: the rank of the free part (number of independent holes)
    torsion: list of integers d_i > 1 (torsion coefficients)
    """
    dim: int
    betti: int
    torsion: List[int] = field(default_factory=list)

    def is_trivial(self):
        return self.betti == 0 and not self.torsion

    def __str__(self):
        parts = []
        if self.betti == 1:
            parts.append('ℤ')
        elif self.betti > 1:
            parts.append(f'ℤ^{self.betti}')
        for d in self.torsion:
            parts.append(f'ℤ/{d}ℤ')
        return ' ⊕ '.join(parts) if parts else '0'

    def __repr__(self):
        return f'H_{self.dim} = {self}'


def homology(K, n):
    """Compute H_n of simplicial complex K."""
    simplices_n = K.simplices(n)
    k_n = len(simplices_n)

    if k_n == 0:
        return HomologyGroup(dim=n, betti=0)

    d_n = boundary_matrix(K, n)
    r_n = matrix_rank(d_n) if d_n else 0

    d_np1 = boundary_matrix(K, n + 1)
    if d_np1:
        factors = smith_factors(d_np1)
        r_np1 = len(factors)
        torsion = [d for d in factors if d > 1]
    else:
        r_np1 = 0
        torsion = []

    betti = k_n - r_n - r_np1
    return HomologyGroup(dim=n, betti=betti, torsion=torsion)


def all_homology(K):
    """Compute H_n for all n from 0 to dimension of K."""
    return [homology(K, n) for n in range(K.dimension + 1)]


def euler_characteristic_from_homology(K):
    """χ(K) = Σ (-1)^n β_n. Should match K.euler_characteristic()."""
    groups = all_homology(K)
    return sum((-1)**g.dim * g.betti for g in groups)


def describe(K, name=None):
    """Print a readable homology summary for K."""
    label = name or repr(K)
    print(f"\n{'─' * 50}")
    print(f"  {label}")
    print(f"{'─' * 50}")
    print(f"  dim = {K.dimension}, χ = {K.euler_characteristic()}")
    for g in all_homology(K):
        marker = '  ' if g.is_trivial() else '→ '
        print(f"  {marker}H_{g.dim} = {g}")
    print()

File: 08/haskell/Homology.hs
module Homology
    ( HomologyGroup(..)
    , homology
    , allHomology
    , eulerFromBetti
    , describe
    ) where

import Simplicial
import Boundary
import Smith

data HomologyGroup = HomologyGroup
    { hDim    :: Int
    , hBetti  :: Int      -- rank of free part
    , hTorsion :: [Int]   -- torsion coefficients > 1
    } deriving (Eq)

-- | H_n = ℤ^β ⊕ ℤ/d_1ℤ ⊕ ... ⊕ ℤ/d_kℤ
instance Show HomologyGroup where
    show (HomologyGroup n b ts) =
        "H_" ++ show n ++ " = " ++ showGroup b ts

showGroup :: Int -> [Int] -> String
showGroup 0 []  = "0"
showGroup b ts  = intercalate " ⊕ " parts
  where
    parts = freePart ++ map torsionPart ts
    freePart
        | b == 0    = []
        | b == 1    = ["ℤ"]
        | otherwise = ["ℤ^" ++ show b]
    torsionPart d = "ℤ/" ++ show d ++ "ℤ"

    intercalate _ []     = ""
    intercalate _ [x]    = x
    intercalate sep (x:xs) = x ++ sep ++ intercalate sep xs

-- | Compute H_n of a simplicial complex.
homology :: SimplicialComplex -> Int -> HomologyGroup
homology k n =
    let kn    = length (simplicesOfDim k n)
        dn    = boundaryMatrix k n
        rn    = matrixRank dn
        dnp1  = boundaryMatrix k (n + 1)
        (rnp1, torsion) =
            if null dnp1
            then (0, [])
            else let fs = smithFactors dnp1
                 in (length fs, filter (> 1) fs)
        betti = kn - rn - rnp1
    in HomologyGroup n betti torsion

-- | Compute H_n for n = 0, 1, ..., dim(K).
allHomology :: SimplicialComplex -> [HomologyGroup]
allHomology k = map (homology k) [0 .. dimension k]

-- | χ = Σ (-1)^n β_n.
eulerFromBetti :: SimplicialComplex -> Int
eulerFromBetti k =
    sum [(-1)^n * hBetti g | (n, g) <- zip [0..] (allHomology k)]

-- | Print a readable homology summary.
describe :: String -> SimplicialComplex -> IO ()
describe name k = do
    putStrLn $ replicate 50 '─'
    putStrLn $ "  " ++ name
    putStrLn $ replicate 50 '─'
    putStrLn $ "  dim = " ++ show (dimension k)
             ++ ", χ = " ++ show (eulerCharacteristic k)
    mapM_ printGroup (allHomology k)
    putStrLn ""
  where
    printGroup g@(HomologyGroup _ b ts)
        | b == 0 && null ts = putStrLn $ "     " ++ show g
        | otherwise          = putStrLn $ "  →  " ++ show g
